blob/drivertest/drivertest.go

```diff
@@ -1821,6 +1825,75 @@ func testDelete(t *testing.T, newHarness HarnessMaker) {
 	})
 }
 
+// testConcurrentWriteAndRead tests that concurrent writing and reading of
+// keys into a blob works.
+func testConcurrentWriteAndRead(t *testing.T, newHarness HarnessMaker) {
+	ctx := context.Background()
+	h, err := newHarness(ctx, t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+	drv, err := h.MakeDriver(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	b := blob.NewBucket(drv)
+	defer b.Close()
+
+	// Prepare data. Each blob dataSize bytes equal to its numeric key.
+	const numKeys = 20
+	const dataSize = 4 * 1024
+	keyData := make(map[int][]byte)
+	for k := 0; k < numKeys; k++ {
+		data := make([]byte, dataSize)
+		for i := 0; i < dataSize; i++ {
+			data[i] = byte(k)
+		}
+		keyData[k] = data
+	}
+
+	blobName := func(k int) string {
+		return fmt.Sprintf("key%d", k)
+	}
+
+	var wg sync.WaitGroup
+
+	// Write all blobs concurrently.
+	for k := 0; k < numKeys; k++ {
+		wg.Add(1)
+		go func(key int) {
+			if err := b.WriteAll(ctx, blobName(key), keyData[key], nil); err != nil {
+				t.Fatal(err)
+			}
+			wg.Done()
+		}(k)
+	}
+	wg.Wait()
+
+	// Read all blobs concurrently and verify that they contain the expected data.
+	for k := 0; k < numKeys; k++ {
+		wg.Add(1)
+		go func(key int) {
+			buf, err := b.ReadAll(ctx, blobName(key))
+			if err != nil {
+				t.Fatal(err)
+			}
+			if len(buf) != dataSize {
+				t.Errorf("data size for key %d, got %d want %d", key, len(buf), dataSize)
+			}
+			for i := 0; i < len(buf); i++ {
```

Comment:
Could also use `bytes.Equal(buf, keyData[k])`.

Reply:
Yes, but it would make error reporting much harder. Emitting huge byte slices isn't too helpful, whereas here you get a helpful error message. It actually saved my bacon once when I got the sync wrong and one key got data from another, was very quick to see what the problem is. Seems worth it for the cost of 4 lines of trivial code?

Reply:
Up to you, but IMHO the error could just emit the key number, not the index. It seems unlikely that some of the bytes are going to be right and others wrong.

Reply:
OK, I checked and we use `bytes.Equal` a bunch in tests, so will align to that. Maybe we should write some of our own testing utils to makes this nicer (or wait for generics)